src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -106,41 +404,319 @@ public function request(string $method, string $url, array $options = []): Respo
         return MockResponse::fromRequest($method, $url, $options, $response);
     }
 
-    public function stream(ResponseInterface|iterable $responses, ?float $timeout = null): ResponseStreamInterface
+    /**
+     * Returns a hash representing the request details.
+     */
+    private function getRequestHash(string $method, string $url): string
     {
-        if ($responses instanceof ResponseInterface) {
-            $responses = [$responses];
-        }
+        return hash('xxh3', $method.$url);
+    }
 
-        $mockResponses = [];
-        $clientResponses = [];
+    /**
+     * Generates a unique metadata key based on the request hash and varying headers.
+     *
+     * @param string                         $requestHash       a hash representing the request details
+     * @param array<string, string|string[]> $normalizedHeaders normalized headers of the request
+     * @param string[]                       $varyFields        headers to consider for building the variant key
+     *
+     * @return string the metadata key composed of the request hash and variant key
+     */
+    private function getMetadataKey(string $requestHash, array $normalizedHeaders, array $varyFields): string
+    {
+        $variantKey = hash(
+            'xxh3',
+            [] === $varyFields
+                ? 'NO-VARY'
+                : $this->buildVariantKey($normalizedHeaders, $varyFields)
+        );
```

Comment:
not sure we need a special case for the empty-vary state:
```suggestion
        $variantKey = hash('xxh3', $this->buildVariantKey($normalizedHeaders, $varyFields));
```